File: apps/web/lib/dossier.ts
export type ClaimStatus =
  | "configuration"
  | "exact"
  | "rounded"
  | "mismatch"
  | "declared_table"
  | "declared_figure"
  | "declared_equation"
  | "declared_theorem"
  | "declared_algorithm"
  | "citation"
  | "untraced";

export interface AnchorRect {
  x0: number;
  y0: number;
  x1: number;
  y1: number;
}

export interface AnchorJson {
  page: number;
  pageWidth: number;
  pageHeight: number;
  precision: string;
  rects: AnchorRect[];
}

export interface Claim {
  value: string;
  section: string;
  sentence: string;
  status: ClaimStatus;
  table: string | null;
  caption: string | null;
  row: string | null;
  column: string | null;
  cell: string | null;
  /** Which label the paper points at, when the evidence is declared rather than verified. */
  evidenceLabel: string | null;
  /** Where the claim is stated in the PDF; null when it could not be located. */
  anchor: AnchorJson | null;
  /** Where its evidence sits — the cell if it is unambiguous, else the caption. */
  evidenceAnchor: AnchorJson | null;
}

export interface TableCellData {
  text: string;
  value: string | null;
  emphasised: boolean;
  block: number;
}

export interface TableData {
  name: string;
  caption: string;
  label: string;
  numericCells: number;
  supports: number;
  rows: TableCellData[][];
  anchor: AnchorJson | null;
}

export interface EquationData {
  ordinal: number;
  latex: string;
  environment: string;
  labels: string[];
}

export interface SectionData {
  kind: string;
  title: string;
  chars: number;
  numbers: number;
  citations: number;
  /** Where the section begins on the page, so the outline can go there. */
  anchor?: AnchorJson | null;
}

export type NumberKind = "result" | "configuration" | "reference" | "structural";

/** One measurement from anywhere in the paper, traced the way a headline claim is. */
export interface IndexedNumber {
  value: string;
  kind: NumberKind;
  section: string;
  sentence: string;
  status: ClaimStatus | "configuration";
  table: string | null;
  row: string | null;
  cell: string | null;
  anchor: AnchorJson | null;
}

export interface Reference {
  key: string;
  raw: string;
  authors: string;
  title: string;
  venue: string;
  year: number | null;
  /** Present for references we could follow to check what they actually reported. */
  arxivId: string | null;
  doi: string | null;
}

export type BaselineOutcome = "confirmed" | "not_found" | "unavailable";

/** One figure a paper attributes to another paper, and whether that paper reports it. */
export interface BaselineCheck {
  table: string;
  row: string;
  column: string;
  value: string;
  citationKey: string;
  arxivId: string | null;
  citedTitle: string;
  outcome: BaselineOutcome;
  note: string;
}

/** What a paper does with a work it cites. Read off the prose, not inferred. */
export type Stance = "inherits" | "extends" | "contests" | "compares" | "background";

export interface LineageEdge {
  key: string;
  stance: Stance;
  /** The exact words in the paper that decided the stance. */
  cue: string;
  sentence: string;
  section: string;
  sectionKind: string;
  title: string;
  authors: string;
  year: number | null;
  arxivId: string | null;
  /** Whether the cited paper is in the library, so the edge can be walked. */
  inCorpus: boolean;
  anchor: AnchorJson | null;
}

export interface LineageTally {
  inherits: number;
  extends: number;
  contests: number;
  compares: number;
  background: number;
  traversable: number;
}

/** What sort of thing the paper is taking for granted. */
export type AssumptionKind =
  | "stated"
  | "simplifying"
  | "conventional"
  | "conjectural"
  | "conditional";

/**
 * What the paper offers for it, in the same sentence. Named for what is observable:
 * a citation in the sentence is a citation in the sentence, and whether it actually
 * establishes the assumption is a judgement the reader makes, not one we claim.
 */
export type AssumptionSupport = "cited" | "shown" | "bare";

export interface Assumption {
  sentence: string;
  anchorText: string;
  kind: AssumptionKind;
  support: AssumptionSupport;
  cue: string;
  section: string;
  sectionKind: string;
  cites: string[];
  refs: string[];
  anchor: AnchorJson | null;
}

/**
 * How often the cue readings are right, measured on a hand-labelled sample.
 *
 * Written by `keystone stance-eval` from the labels themselves, so the number on the
 * site cannot drift from the file it came from. The site's whole claim is that these
 * readings are checkable; how often they are correct belongs on the site.
 */
export interface Accuracy {
  labelled: number;
  heldOut: number;
  /** Accuracy on rows drawn *after* the rules were last changed. The honest figure. */
  heldOutAccuracy: number | null;
  /** 95% Wilson interval on that figure. Thirty rows cannot pin it down further. */
  heldOutInterval: [number, number];
  /** The metric published citation-intent work reports; accuracy flatters the
   * majority class, which is over half of every corpus in that literature. */
  macroF1: number;
  baselineMacroF1: number;
  accuracy: number;
  baselineAccuracy: number;
  majorityAccuracy: number;
  classes: Record<
    string,
    { gold: number; predicted: number; precision: number | null; recall: number | null }
  >;
  /** The same rules against labels written by other people, on papers outside the
   * library. Everything above describes rules measured on the corpus they were
   * written against; this is the only part that says how far they travel. */
  external?: External[];
}

/**
 * The cue rules scored against a public citation-intent corpus.
 *
 * `coverage` and `spokenPrecision` are the pair the site quotes. `macroF1` is what
 * published work reports, and it is included because leaving it out would be the
 * convenient choice: it punishes silence exactly as hard as error, and Keystone is
 * built to be silent, so the figure is low and says so.
 */
export interface External {
  corpus: string;
  instances: number;
  /** Share of citations a stance was reported for at all. */
  coverage: number;
  /** How many that is, in citations. `coverage` is this over `instances`. */
  spoken: number;
  /** How often those were right. Excludes the silent class by construction. */
  spokenPrecision: number;
  spokenInterval: [number, number];
  macroF1: number;
  /** The best macro-F1 reachable here: ACL-ARC has two classes Keystone cannot say. */
  macroF1Ceiling: number;
  unreachable: string[];
}

export interface LibraryGraph {
  nodes: {
    id: string;
    title: string;
    inherits: number;
    contests: number;
    bare: number;
  }[];
  edges: {
    from: string;
    to: string;
    stance: Stance;
    cue: string;
    sentence: string;
    section: string;
    anchor: AnchorJson | null;
  }[];
}

export interface Dossier {
  id: string;
  title: string;
  pdfUrl: string;
  /** What arXiv files this paper under. Absent for papers whose metadata failed. */
  arxiv?: ArxivRecord | null;
  /** The repository the paper itself names, if it names one. Never inferred. */
  codeUrl?: string;
  keystone: {
    table: string;
    caption: string;
    supported: number;
    share: number;
    summary: string;
  } | null;
  coverage: {
    claims: number;
    supported: number;
    declared: number;
    unsupported: number;
    mismatched: number;
    rate: number;
  };
  claims: Claim[];
  numbers: IndexedNumber[];
  references: Reference[];
  baselines: BaselineCheck[];
  tables: TableData[];
  equations: EquationData[];
  /** The paper's own \newcommand definitions, for rendering its notation. */
  macros: Record<string, string>;
  sections: SectionData[];
  lineage: {
    tally: LineageTally;
    /** The one work this paper would not stand without. */
    foundation: LineageEdge | null;
    edges: LineageEdge[];
  };
  assumptions: Assumption[];
  assumptionTally: { total: number; cited: number; shown: number; bare: number };
  findings: {
    check_id: string;
    title: string;
    severity: string;
    confidence: string;
    explanation: string;
    arithmetic: Record<string, unknown>;
  }[];
}

export interface IndexEntry {
  id: string;
  title: string;
  coverage: Dossier["coverage"];
  keystone: Dossier["keystone"];
  findings: number;
  /**
   * How much the paper has in it beyond its headline numbers. Four of nine papers
   * state no numbers up front, and without this the library row for one of them says
   * only "no numeric claims" — which reads as a failure rather than as a paper that
   * argues in prose and still has 58 references and 10 tables to look at.
   */
  density?: {
    numbers: number;
    tables: number;
    equations: number;
    references: number;
  };
  /** The paper's own full title, used to resolve citations across the library. */
  fullTitle?: string;
  lineage?: LineageTally;
  assumptions?: { total: number; cited: number; shown: number; bare: number };
  arxiv?: ArxivRecord | null;
}

/**
 * What arXiv says a paper is, as opposed to what this project infers about it.
 *
 * The library groups by `primary`, and the reason it is fetched rather than derived
 * is the same reason every quote is verbatim: "the authors filed this under cs.CL" is
 * checkable on the abstract page, and "this looks like an NLP paper" is my opinion.
 */
export interface ArxivRecord {
  /** Primary category code, e.g. "cs.CL". */
  primary: string;
  /** arXiv's own name for it, not an abbreviation of mine. */
  primaryName: string;
  /** Every category, primary first — most papers are cross-listed. */
  categories: string[];
  /** ISO date of version 1. */
  published: string;
  /** Who wrote it, in arXiv's order. */
  authors?: string[];
}

/** Papers with no fetched record, grouped under one honest heading. */
export const UNCATEGORISED = "uncategorised";

export function categoryOf(paper: IndexEntry): string {
  return paper.arxiv?.primary || UNCATEGORISED;
}

export function categoryLabel(paper: IndexEntry): string {
  return paper.arxiv?.primaryName || "Not yet categorised";
}

/** Inheriting a method makes the cited paper's correctness a precondition of yours. */
export const isLoadBearing = (stance: Stance) =>
  stance === "inherits" || stance === "extends";

export const STANCE_VERB: Record<Stance, string> = {
  inherits: "adopts",
  extends: "builds on",
  contests: "argues with",
  compares: "measures against",
  background: "mentions",
};

/** Verified by arithmetic: the number was found where it should be. */
export const isSupported = (status: ClaimStatus) =>
  status === "exact" || status === "rounded";

/**
 * The paper names its evidence, but of a kind no arithmetic can confirm — a figure, a
 * derivation, a theorem, another paper. Distinct from unsupported on purpose:
 * "evidenced by Figure 4" and "we could find nothing" are different answers, and
 * collapsing them made every theory and figure-driven result look unevidenced.
 */
export const isDeclared = (status: ClaimStatus) =>
  status.startsWith("declared_") || status === "citation";

const EVIDENCE_WORDS: Partial<Record<ClaimStatus, string>> = {
  declared_table: "a table",
  declared_figure: "a figure",
  declared_equation: "a derivation",
  declared_theorem: "a theorem",
  declared_algorithm: "an algorithm",
  citation: "another paper",
};

export const declaredAs = (status: ClaimStatus) => EVIDENCE_WORDS[status] ?? "evidence";

File: services/core/keystone/dossier.py
"""Assembling a paper's dossier — the object the reader actually sees."""

from __future__ import annotations

from dataclasses import dataclass, replace
from decimal import Decimal
from pathlib import Path
from typing import Any

import keystone.audit.checks.tables  # noqa: F401  (registers the checks)
from keystone.audit.registry import run
from keystone.audit.trace import (
    Coverage,
    Trace,
    all_mentions,
    headline_mentions,
    mismatch_findings,
    trace_all,
)
from keystone.graph.models import Finding, Paper
from keystone.ingest.anchors import Anchor, DocIndex
from keystone.ingest.arxiv_source import load
from keystone.ingest.latex import label_kinds, source_sentences
from keystone.ingest.pdf import Document
from keystone.ingest.sections import extract_sections
from keystone.ingest.tables import build_tables


def _anchor_json(anchor: Anchor | None, pages: dict[int, tuple[float, float]]) -> dict | None:
    """An anchor in a form the reader can draw, with the page box to scale against.

    Rectangles are in PDF points; the viewer renders at whatever width it has. Sending
    the page dimensions alongside lets it scale without having to know anything about
    the document.
    """
    if anchor is None or not anchor.rects:
        return None
    page = anchor.rects[0].page
    width, height = pages.get(page, (612.0, 792.0))
    return {
        "page": page,
        "pageWidth": width,
        "pageHeight": height,
        "precision": anchor.precision,
        "rects": [
            {"x0": r.x0, "y0": r.y0, "x1": r.x1, "y1": r.y1}
            for r in anchor.rects
            if r.page == page
        ],
    }


def _grid(table) -> list[list[dict]]:
    """A table's cells as a dense 2-D grid, ready to render.

    Cells are stored sparsely — blanks are simply absent — so the grid has to be
    filled out before it can be laid out, or every row with an empty cell would shift
    left and the columns would stop lining up with their headers.
    """
    if not table.cells:
        return []
    rows = max(c.row for c in table.cells) + 1
    columns = max(c.column for c in table.cells) + 1
    grid = [[{"text": "", "value": None, "emphasised": False, "block": 0}
             for _ in range(columns)] for _ in range(rows)]
    for cell in table.cells:
        grid[cell.row][cell.column] = {
            "text": cell.raw,
            "value": str(cell.number.value) if cell.number else None,
            "emphasised": cell.is_emphasised,
            "block": cell.block,
        }
    return grid


@dataclass(frozen=True, slots=True)
class Dossier:
    paper: Paper
    coverage: Coverage
    findings: tuple[Finding, ...]
    claim_anchors: tuple[Anchor | None, ...] = ()
    evidence_anchors: tuple[Anchor | None, ...] = ()
    page_sizes: dict[int, tuple[float, float]] = None  # type: ignore[assignment]
    equations: tuple = ()
    numbers: tuple = ()
    number_anchors: tuple = ()
    macros: dict = None  # type: ignore[assignment]
    section_counts: tuple = ()
    table_anchors: dict[int, Anchor | None] = None  # type: ignore[assignment]

    def to_dict(self) -> dict[str, Any]:
        keystone = self.coverage.keystone
        pages = self.page_sizes or {}
        claim_anchors = self.claim_anchors or ((None,) * len(self.coverage.claims))
        table_anchors = self.table_anchors or {}
        # Without a PDF there are no anchors, and zipping against an empty tuple would
        # silently drop every number rather than yielding them unanchored.
        number_anchors = self.number_anchors or ((None,) * len(self.numbers))
        evidence_anchors = self.evidence_anchors or ((None,) * len(self.coverage.claims))
        return {
            "id": self.paper.id,
            "title": self.paper.title,
            "keystone": None if keystone is None else {
                "table": keystone.table.name,
                "caption": keystone.table.caption,
                "supported": keystone.supported,
                "share": round(keystone.share, 3),
                "summary": keystone.summary,
            },
            "coverage": {
                "claims": len(self.coverage.claims),
                "supported": len(self.coverage.supported),
                "declared": len(self.coverage.declared),
                "unsupported": len(self.coverage.unsupported),
                "mismatched": len(self.coverage.mismatched),
                "rate": round(self.coverage.rate, 3),
            },
            "claims": [
                {
                    "value": t.mention.number.raw,
                    "section": str(t.mention.section),
                    "sentence": t.mention.sentence,
                    "status": t.status,
                    "table": t.table.name if t.table else None,
                    "caption": t.table.caption if t.table else None,
                    "row": t.cell.row_header if t.cell else None,
                    "column": t.cell.column_header if t.cell else None,
                    "cell": t.cell.raw if t.cell else None,
                    "evidenceLabel": t.evidence_label,
                    "anchor": _anchor_json(claim_anchors[i], pages),
                    "evidenceAnchor": _anchor_json(evidence_anchors[i], pages),
                }
                for i, t in enumerate(self.coverage.claims)
            ],
            "pdfUrl": f"https://arxiv.org/pdf/{self.paper.id}",
            # Every measurement in the paper, not only the ones in the abstract. The
            # reader used to show 28 of 1,554 numbers across nine papers, which is why
            # four of them had nothing at all to display.
            "numbers": [
                {
                    "value": t.mention.number.raw,
                    "kind": t.mention.kind,
                    "section": str(t.mention.section),
                    "sentence": t.mention.sentence,
                    "status": t.status,
                    "table": t.table.name if t.table else None,
                    "row": t.cell.row_header if t.cell else None,
                    "cell": t.cell.raw if t.cell else None,
                    "anchor": _anchor_json(anchor, pages),
                }
                for t, anchor in zip(self.numbers, number_anchors, strict=True)
            ],
            "tables": [
                {
                    "name": table.name,
                    "caption": table.caption,
                    "label": table.label,
                    "numericCells": len(table.numeric_cells),
                    "supports": sum(
                        1 for t in self.coverage.supported
                        if t.table is not None and t.table.ordinal == table.ordinal
                    ),
                    # The grid itself, so the reader can show the evidence rather than
                    # name it. Cells carry the band they sit in and whether the author
                    # emphasised them, because both are claims the table is making.
                    "rows": _grid(table),
                    "anchor": _anchor_json(table_anchors.get(table.ordinal), pages),
                }
                for table in self.paper.tables
            ],
            "macros": self.macros or {},
            "equations": [
                {
                    "ordinal": equation.ordinal,
                    "latex": equation.latex,
                    "environment": equation.environment,
                    "labels": list(equation.labels),
                }
                for equation in self.equations
            ],
            "sections": [
                {
                    "kind": str(section.kind),
                    "title": section.title,
                    "chars": len(section.text),
                    "numbers": counts.get("numbers", 0),
                    "citations": counts.get("citations", 0),
                }
                for section, counts in self.section_counts
            ],
            "findings": [f.to_dict() for f in self.findings],
        }


def build(
    arxiv_id: str, cache_dir: Path, title: str = "", pdf_path: Path | None = None
) -> Dossier:
    """Ingest a paper and produce everything the reader sees. No model in the loop."""
    document = load(arxiv_id, cache_dir)
    sections = extract_sections(document.text)
    tables = build_tables(document.tables)

    paper = Paper(
        id=arxiv_id,
        title=title,
        sections=sections,
        tables=tables,
        mentions=headline_mentions(sections),
        notes=document.text,
    )
    # Every sentence in the paper, so a headline number can be followed to wherever
    # the body restates it and says what it rests on.
    kinds = label_kinds(document.text)
    body = tuple(source_sentences(document.text))
    coverage = trace_all(paper.mentions, paper.tables, kinds, body)

    # The full index: every measurement the paper states, traced the same way the
    # headline claims are. Configuration values are carried too — a reader wants to
    # find the batch size as much as the accuracy — but they are not traced, because
    # a hyperparameter is not a claim that needs evidence.
    # Results are traced; configuration is indexed but not. A batch size does not
    # need evidence, and giving it a trace status ("rests on a derivation") is noise
    # dressed as analysis.
    measured = all_mentions(sections)
    numbers = (
        trace_all(
            tuple(m for m in measured if m.kind == "result"), paper.tables, kinds, body
        ).traces
        + tuple(
            Trace(mention=m, status="configuration")
            for m in measured
            if m.kind == "configuration"
        )
    )
    findings = tuple(run(paper)) + tuple(mismatch_findings(coverage))

    claim_anchors: tuple[Anchor | None, ...] = ()
    evidence_anchors: tuple[Anchor | None, ...] = ()
    table_anchors: dict[int, Anchor | None] = {}
    number_anchors: tuple = ()
    page_sizes: dict[int, tuple[float, float]] = {}

    if pdf_path is not None and pdf_path.exists():
        pdf = Document.open(pdf_path)
        index = DocIndex(pdf)
        page_sizes = {p.number: (p.width, p.height) for p in pdf.pages}
        claim_anchors = tuple(
            index.locate(t.mention.sentence).anchor for t in coverage.claims
        )
        evidence_anchors = tuple(_evidence_anchor(index, t) for t in coverage.claims)
        # Sentences repeat, so locate each one once rather than once per number in it.
        located: dict[str, Anchor | None] = {}
        for trace in numbers:
            sentence = trace.mention.sentence
            if sentence not in located:
                located[sentence] = index.locate(sentence).anchor
        number_anchors = tuple(located[t.mention.sentence] for t in numbers)
        # Tables are located by their caption, which is long enough to match reliably
        # and puts the reader at the right place on the page.
        table_anchors = {
            table.ordinal: index.locate(table.caption).anchor if table.caption else None
            for table in tables
        }

    return Dossier(
        paper=paper,
        coverage=coverage,
        findings=findings,
        claim_anchors=claim_anchors,
        evidence_anchors=evidence_anchors,
        page_sizes=page_sizes,
        equations=tuple(document.equations),
        numbers=numbers,
        number_anchors=number_anchors,
        macros=document.macros,
        section_counts=_section_counts(sections),
        table_anchors=table_anchors,
    )


def _section_counts(sections: tuple) -> tuple:
    """Per-section density, so the outline can show the shape of a paper's evidence."""
    from keystone.audit.numbers import find_numbers
    from keystone.ingest.latex import source_sentences

    out = []
    for section in sections:
        sentences = source_sentences(section.source or section.text)
        out.append((
            section,
            {
                "numbers": sum(len(find_numbers(s.text)) for s in sentences),
                "citations": sum(len(s.cites) for s in sentences),
            },
        ))
    return tuple(out)


def _evidence_anchor(index: DocIndex, trace) -> Anchor | None:
    """Where in the PDF the evidence for a claim lives.

    Preference order matters. A caption locates reliably and takes the reader to the
    right table, but pointing at the *cell* is what makes the claim checkable at a
    glance. So the cell is tried first — qualified by its row label, since a bare
    "28.4" occurs several times on a page and an ambiguous match is refused — and the
    caption is the fallback rather than the goal.
    """
    if trace.table is None or not trace.table.caption:
        return None

    caption = index.locate(trace.table.caption)
    if caption.anchor is None:
        return None
    page = caption.anchor.rects[0].page if caption.anchor.rects else None
    if page is None:
        return caption.anchor

    cell = trace.cell
    if cell is not None:
        for needle in (f"{cell.row_header} {cell.raw}", cell.raw):
            if not needle.strip():
                continue
            located = index.locate_on_page(needle, page)
            if located is not None:
                return located

    return caption.anchor
